models/public/wavernn/0001-Added-batch-norm-fusing-to-conv-layers.patch

```diff
@@ -0,0 +1,171 @@
++    return fused_conv
++
+ class ResBlock(nn.Module):
+     def __init__(self, dims):
+         super().__init__()
+@@ -17,16 +39,28 @@ class ResBlock(nn.Module):
+         self.conv2 = nn.Conv1d(dims, dims, kernel_size=1, bias=False)
+         self.batch_norm1 = nn.BatchNorm1d(dims)
+         self.batch_norm2 = nn.BatchNorm1d(dims)
++        self.fused = False
+ 
+     def forward(self, x):
+         residual = x
+-        x = self.conv1(x)
+-        x = self.batch_norm1(x)
+-        x = F.relu(x)
+-        x = self.conv2(x)
+-        x = self.batch_norm2(x)
++
++        if self.fused:
++            x = self.conv1_f(x)
++            x = F.relu(x)
++            x = self.conv2_f(x)
++        else:
++            x = self.conv1(x)
++            x = self.batch_norm1(x)
++            x = F.relu(x)
++            x = self.conv2(x)
++            x = self.batch_norm2(x)
+         return x + residual
+ 
++    def fuse(self):
++        self.conv1_f = fuse(self.conv1, self.batch_norm1)
++        self.conv2_f = fuse(self.conv2, self.batch_norm2)
++        self.fused = True
++
+ 
+ class MelResNet(nn.Module):
+     def __init__(self, res_blocks, in_dims, compute_dims, res_out_dims, pad):
+@@ -38,15 +72,26 @@ class MelResNet(nn.Module):
+         for i in range(res_blocks):
+             self.layers.append(ResBlock(compute_dims))
+         self.conv_out = nn.Conv1d(compute_dims, res_out_dims, kernel_size=1)
++        self.fused = False
+ 
+     def forward(self, x):
+-        x = self.conv_in(x)
+-        x = self.batch_norm(x)
++        if self.fused:
++            x = self.conv_in_f(x)
++        else:
++            x = self.conv_in(x)
++            x = self.batch_norm(x)
+         x = F.relu(x)
+         for f in self.layers: x = f(x)
+         x = self.conv_out(x)
+         return x
+ 
++    def fuse(self):
++        self.conv_in_f = fuse(self.conv_in, self.batch_norm)
++        self.fused = True
++
++        for l in self.layers:
++            l.fuse()
++
+ 
+ class Stretch2d(nn.Module):
+     def __init__(self, x_scale, y_scale):
+@@ -61,6 +106,16 @@ class Stretch2d(nn.Module):
+         return x.view(b, c, h * self.y_scale, w * self.x_scale)
+ 
+ 
++class StretchUpsample(nn.Module):
++    def __init__(self, x_scale):
++        super().__init__()
++        self.upsample = nn.Upsample(scale_factor=x_scale)
++
++    def forward(self, x):
++        print('USHAPE: ', x.shape)
++        return self.upsample(x)
++
++
+ class UpsampleNetwork(nn.Module):
+     def __init__(self, feat_dims, upsample_scales, compute_dims,
+                  res_blocks, res_out_dims, pad):
+@@ -68,12 +123,12 @@ class UpsampleNetwork(nn.Module):
+         total_scale = np.cumproduct(upsample_scales)[-1]
+         self.indent = pad * total_scale
+         self.resnet = MelResNet(res_blocks, feat_dims, compute_dims, res_out_dims, pad)
+-        self.resnet_stretch = Stretch2d(total_scale, 1)
++        self.resnet_stretch = StretchUpsample((1,total_scale)) #Stretch2d(total_scale, 1)
```

Comment:
You should remove the comment. This is a diff; you can already see the old version of the code.